DOCSITE: Reorganise tutorials and add inline solutions (#231)
seL4/docs: Tutorials/seL4Kernel/setting-up.md

---
toc: true
layout: tutorial
SPDX-License-Identifier: CC-BY-SA-4.0
SPDX-FileCopyrightText: 2024 seL4 Project a Series of LF Projects, LLC.
---

# Setting up your machine
**Overview**
- Set up your machine - install dependencies required to run seL4``
- Run seL4test on a simulator

> - Gain awareness of terminology used for seL4

Reviewer: These should either be links or a paragraph. "Overview" should be a heading.

Author: I've removed this, since the contents list on the right provides links to each of the sections that are on the page.

## Docker

> To compile and use seL4, it is recommended that you use Docker to isolate the dependencies from your machine.

Reviewer: If we are recommending docker, we don't really need the rest of the page apart from `repo` (although I don't know if the virtual machine tutorials actually work under docker). We either should remove the docker section or add some text that makes clear that docker and the rest are two alternatives, and that you need only one of them.

Author: I actually removed everything but the repo and docker sections, because there is a whole other page on dependencies https://docs.sel4.systems/projects/buildsystem/host-dependencies.html, should someone need them.

I couldn't get of the virtual machine tutorials working on my silicon mac (but I don't think it's related to Docker at all).

Adding issues here for completeness:

"CAmkES VM: Adding a Linux Guest" uses KVM and I would have needed an alternative for mac. 

"CAmkES VM: Cross VM Connectors" breaks using `ninja`.

```
sh.ErrorReturnCode_1: 

  RAN: /usr/bin/cmake -G Ninja -DTUT_BOARD=pc -DTUT_ARCH=x86_64 -DFORCE_IOMMU=ON -DTUTORIAL_DIR=camkes-vm-crossvmke7jqji2 -C ../projec

  STDOUT:

  STDERR:
```

Should I add a note to these tutorials saying the instructions are for Linux only?

Reviewer: Yes, that would be good. We've already had some questions from people who were trying to use other platforms.

These instructions are intended for Ubuntu LTS versions 20.04 and 22.04.

To begin, you will need at least these two programs:

 * make (`sudo apt install make`)
 * docker (See [here](https://get.docker.com) or [here](https://docs.docker.com/engine/installation) for installation instructions)

For convenience, add your account to the docker group:

```bash
sudo usermod -aG docker $(whoami)
```

<details markdown='1'>
  <summary style="display:list-item">More on Docker</summary>
  <br>

  **Available images**

  All the prebuilt docker images are available on [DockerHub here](https://hub.docker.com/u/trustworthysystems)

  These images are used by the Trustworthy Systems Continuous Integration (CI) software, and so represent a standard software setup we use.
  The CI software always uses the `latest` docker image, but images are also tagged with the date they were built.

  **More information**

  You can find the dockerfiles and supporting Makefile [here](https://github.com/seL4/seL4-CAmkES-L4v-dockerfiles)

</details>

## Google's Repo tool

> The primary way of obtaining and managing seL4 project source is through the use of Google's Repo tool.

Reviewer: ```suggestion
The primary way of obtaining and managing seL4 project sources is through the use of Google's Repo tool.
```

To install run:
```sh
  sudo apt-get update
  sudo apt-get install repo
```

<details markdown='1'>

> <summary style="display:list-item">More on Repo</summary>

Reviewer: Those details sections are nice!

```sh
sudo apt-get install gcc-arm-linux-gnueabi g++-arm-linux-gnueabi
sudo apt-get install gcc-aarch64-linux-gnu g++-aarch64-linux-gnu
```

(you can install the hardware floating point versions as well if you wish)

```sh
sudo apt-get install gcc-arm-linux-gnueabihf g++-arm-linux-gnueabihf
```

### Cross-compiling for RISC-V targets

To build for RISC-V targets you will need a cross compiler:

{% include risc-v.md %}

<details markdown='1'>
<summary style="display:list-item">More on host dependencies</summary>
<br>
[A detailed guide on host dependencies](../projects/buildsystem/host-dependencies)
</details>

## Python Dependencies

Regardless of your Linux distribution, python dependencies are required to build seL4, the manual and its proofs. To install you can run:

```sh
pip3 install --user setuptools
pip3 install --user sel4-deps
```

(Some distributions use `pip` for python3, others use `pip3`. Use the Python 3 version for your distribution.)

## Getting a build environment
To get a running build environment for seL4 and Camkes, run:

```bash
git clone https://github.com/seL4/seL4-CAmkES-L4v-dockerfiles.git
cd seL4-CAmkES-L4v-dockerfiles

> make user

Reviewer: These are now docker instructions again. They should be in the docker section.

```

This will give you a terminal inside a container that has all the relevant tools to build, simulate, and test seL4 & Camkes programs.

The first time you run this, docker will fetch the relevant images, which may take a while.

The last line will say something like:

```
Hello, welcome to the seL4/CAmkES/L4v docker build environment
```

## Mapping a container
To run the container from other directories (e.g. starting a container for the [Hello World](hello-world) tutorial, which we'll do next), you can setup a bash alias such as this:

```bash
echo $'alias container=\'make -C /<path>/<to>/seL4-CAmkES-L4v-dockerfiles user HOST_DIR=$(pwd)\'' >> ~/.bashrc
# now open a new terminal, or run `source ~/.bashrc`
```

Replace `/<path>/<to>/` to match where you cloned the git repo of the docker files.

*Reminder:* Include the absolute path to your `seL4-CAmkES-L4v-dockerfiles` folder, e.g.

```bash

> echo $'alias container=\'make -C //home/jblogs/seL4-CAmkES-L4v-dockerfiles user HOST_DIR=$(pwd)\'' >> ~/.bashrc

Reviewer: `~/` instead of `/home/jblogs/` should work fine

# now open a new terminal, or run `source ~/.bashrc`

```

This then allows you to run `container` from any directory.

> *Reminder:* Restart Ubuntu or run a new terminal for the changes in `.bashrc` to take effect.

Reviewer: ```suggestion
*Reminder:* start a new terminal for the changes in `.bashrc` to take effect or run `source ~/.bashrc`.
```

## An example workflow

A good workflow is to run two terminals:

>  - Terminal A is just a normal terminal, and is used for git operations, editing (e.g., vim, emacs), and other normal operations.

Reviewer: ```suggestion
 - Terminal A is just a normal terminal, and is used for git operations, editing (e.g., vim, emacs, vscode), and other normal operations.
```
